plugins/methodinfo/src/main/java/org/qbicc/plugin/methodinfo/MethodInformation.java

```diff
@@ -0,0 +1,503 @@
+        return new LLVMGenerator(isPie ? 2 : 0, isPie ? 2 : 0).processProgramModule(ctxt, programModule);
+    }
+
+    public MethodInformation(boolean isPie) {
+        this.isPie = isPie;
+    }
+
+    private void compileMethodDataModule(CompilationContext context, Path modulePath) {
+        LLVMCompileStage compileStage = new LLVMCompileStage(isPie);
+        compileStage.compileModule(context, modulePath);
+    }
+
+    @Override
+    public void accept(CompilationContext context) {
+        createMethodData(context);
+        Path modulePath = emitMethodData(context);
+        if (modulePath != null) {
+            compileMethodDataModule(context, modulePath);
+        }
+    }
+
+    private static class StackMapRecord implements Comparable {
+        private final int objectFileIndex;
+        private final int offset;
+        private final long statepoindId;
+
+        StackMapRecord(final int objectFileIndex, final int offset, final long statepoindId) {
+            this.objectFileIndex = objectFileIndex;
+            this.offset = offset;
+            this.statepoindId = statepoindId;
+        }
+
+        long getStatepoindId() { return statepoindId; }
+        int getOffset() { return offset; }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(objectFileIndex, statepoindId, offset);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            return equals((StackMapRecord)o);
+        }
+
+        public boolean equals(StackMapRecord record) {
+            return objectFileIndex == record.objectFileIndex
+                && offset == record.offset;
+        }
+
+        @Override
+        public int compareTo(Object o) {
+            StackMapRecord other = (StackMapRecord) o;
+            if (equals(other)) {
+                return 0;
+            }
+            if (objectFileIndex < other.objectFileIndex) {
+                return -1;
+            } else if (objectFileIndex > other.objectFileIndex) {
+                return 1;
+            } else if (offset < other.offset) { // if same objectFileIndex then order by offset
+                return -1;
+            } else {
+                return 1;
+            }
+        }
+    }
+
+    private final class StackMapRecordCollector {
+        CompilationContext context;
+        StackMapRecordCollector(CompilationContext context) {
+            this.context = context;
+        }
+
+        public List<StackMapRecord> collect() {
+            Linker linker = Linker.get(context);
+            List<StackMapRecord> recordList = new ArrayList<>();
+            ObjectFileProvider objFileProvider = context.getAttachment(Driver.OBJ_PROVIDER_TOOL_KEY);
+            Iterator<Path> objFileIterator = linker.getObjectFilePaths().iterator();
+            final Integer[] index = { 0 };
```

Comment:
Using an `AtomicInteger` here might be cleaner, and certainly less memory-intensive than dealing with `Integer`.  Or else make it an `int[]` rather than `Integer[]`.

Reply:
Changed it to `int[]`.